In [1]:
import pandas as pd

### V-dem variables for 2019 year (commented out due to file size)

In [2]:
#vdem_whole = pd.read_csv('vdem_march_2020.csv', low_memory=False)
# vdem = vdem_whole.loc[vdem_whole.year >= 2019]
# vdem.shape

In [3]:
# # there is no data for 2019:
# #'e_wbgi_rqe' (regulatory quality)
# #'e_peedgini' (education variable)
# # 'e_cow_exports', 'e_cow_imports' (export and import)
# vars =['country_name', 'country_text_id', 'year', 'e_regiongeo', 'e_regionpol', 'v2x_rule', 'v2xcl_acjst', 'v2xcl_prpty', 'v2x_polyarchy', 'e_peaveduc']
# for i in vars:
#     print(i, '\n')
#     print(vdem[i].isna().sum())

In [4]:
# vdem = vdem[vars]
# vdem.rename(columns={'e_regiongeo': 'region_geo', 'e_regionpol': 'region_pol', "v2x_polyarchy": 'polyarchy',  'v2x_rule': 'rule', 'v2xcl_acjst': 'justice', 'v2xcl_prpty': 'property', 'e_peaveduc': 'education'}, inplace = True)
# vdem = vdem.sort_index(axis=1, ascending=False)
# vdem.isna().sum()
# vdem.to_csv('vdem_2019.csv')

In [5]:
vdem = pd.read_csv('vdem_2019.csv')
vdem.drop(columns=['Unnamed: 0'], inplace=True)

In [6]:
vdem.head()

country_name  rule_2017  rule_2018  rule_2019  region_pol  region_geo  \
0  Afghanistan      0.219      0.254      0.341           8          14   
1      Albania      0.514      0.543      0.596           1           3   
2      Algeria      0.204      0.200      0.217           3           5   
3       Angola      0.306      0.400      0.406           4           7   
4    Argentina      0.689      0.732      0.757           2          18   

   property_2017  property_2018  property_2019  polyarchy_2017  \
0          0.417          0.436          0.423           0.369   
1          0.755          0.753          0.737           0.563   
2          0.723          0.729          0.611           0.306   
3          0.614          0.580          0.603           0.316   
4          0.872          0.868          0.824           0.745   

   polyarchy_2018  polyarchy_2019  justice_2017  justice_2018  justice_2019  \
0           0.368           0.347         0.205         0.174         0.198   
1           0.499           0.481         0.875         0.872         0.936   
2           0.320           0.300         0.615         0.643         0.500   
3           0.354           0.386         0.451         0.451         0.487   
4           0.834           0.812         0.763         0.819         0.707   

   education_2017  education_2018  education_2019 country_text_id  
0           3.813           3.813           3.813             AFG  
1             NaN             NaN             NaN             ALB  
2           7.300           7.300           7.300             DZA  
3           2.980           2.980           2.980             AGO  
4           8.770           8.770           8.770             ARG

### Covid variables

#### V-dem data

In [7]:
data = pd.read_csv('full_data_V3.csv')

In [8]:
data.head()

country_name        time  emlaw  \
0  Afghanistan  2020-06-17      5   
1      Albania  2020-06-08      3   
2      Algeria  2020-06-22      4   
3       Angola  2020-06-15      1   
4    Argentina  2020-06-09      4   

                                              emname     emstart       emend  \
0                                                NaN         NaN         NaN   
1  Constitution (Articles 100, 170, 174, point 1,...  2020-03-24         NaN   
2  Constitution 1989 (reinst. 1996, rev. 2016), a...  2020-03-21         NaN   
3           Constitution Article 58 and Law n. 17/91  2020-03-27  2020-05-25   
4  Ley 27541 - LEY DE SOLIDARIDAD SOCIAL Y REACTI...  2020-03-20         NaN   

   emlimit emlimitdate  emlawapp emlegapp  ...   panback  pandem_dis  \
0      NaN         NaN       NaN      NaN  ...  0.000000        0.05   
1      1.0  2020-06-23       1.0        1  ...  0.173110        0.15   
2      1.0  2020-04-04       0.0        0  ...  0.123340        0.35   
3      1.0  2020-04-11       1.0        1  ...  0.041671        0.10   
4      1.0  2020-03-31       1.0        1  ...  0.054786        0.05   

   pandem_dis_alt  type1  type2 type3 type4  type4_alt  type5  type6  
0            0.00      0      0     0     1          0      0      0  
1            0.15      0      0     0     3          3      0      0  
2            0.35      0      0     0     3          3      2      0  
3            0.05      0      0     0     1          0      0      1  
4            0.05      0      0     0     0          0      0      1  

[5 rows x 41 columns]

In [9]:
data = data[['country_name','pandem','pandem_dis','panback']]

In [10]:
data.head()

country_name    pandem  pandem_dis   panback
0  Afghanistan  0.058824        0.05  0.000000
1      Albania  0.176471        0.15  0.173110
2      Algeria  0.294118        0.35  0.123340
3       Angola  0.117647        0.10  0.041671
4    Argentina  0.058824        0.05  0.054786

#### Adding continents and sub-region columns to dataframe

https://github.com/lukes/ISO-3166-Countries-with-Regional-Codes/blob/master/all/all.csv

In [11]:
continents = pd.read_csv('all.csv')

In [12]:
continents.head()

name alpha-2 alpha-3  country-code     iso_3166-2   region  \
0     Afghanistan      AF     AFG             4  ISO 3166-2:AF     Asia   
1   Åland Islands      AX     ALA           248  ISO 3166-2:AX   Europe   
2         Albania      AL     ALB             8  ISO 3166-2:AL   Europe   
3         Algeria      DZ     DZA            12  ISO 3166-2:DZ   Africa   
4  American Samoa      AS     ASM            16  ISO 3166-2:AS  Oceania   

        sub-region intermediate-region  region-code  sub-region-code  \
0    Southern Asia                 NaN        142.0             34.0   
1  Northern Europe                 NaN        150.0            154.0   
2  Southern Europe                 NaN        150.0             39.0   
3  Northern Africa                 NaN          2.0             15.0   
4        Polynesia                 NaN          9.0             61.0   

   intermediate-region-code  
0                       NaN  
1                       NaN  
2                       NaN  
3                       NaN  
4                       NaN

In [13]:
continents = continents[['name','alpha-3','region','sub-region']]

In [14]:
continents.rename(columns={"name": "country_name", "alpha-3": "country_txt", "region": "continent"}, inplace=True)

In [15]:
df = pd.merge(data, continents, how='left', on=['country_name'])

In [16]:
df.head()

country_name    pandem  pandem_dis   panback country_txt continent  \
0  Afghanistan  0.058824        0.05  0.000000         AFG      Asia   
1      Albania  0.176471        0.15  0.173110         ALB    Europe   
2      Algeria  0.294118        0.35  0.123340         DZA    Africa   
3       Angola  0.117647        0.10  0.041671         AGO    Africa   
4    Argentina  0.058824        0.05  0.054786         ARG  Americas   

                        sub-region  
0                    Southern Asia  
1                  Southern Europe  
2                  Northern Africa  
3               Sub-Saharan Africa  
4  Latin America and the Caribbean

In [17]:
len(data)

146

In [18]:
len(df)

146

#### Checking which countries names are different in both datasets and then updating their names

In [19]:
df1 = df[df.isna().any(axis=1)]
df1['country_name']

13                              Bolivia
19                        Burma/Myanmar
32                       Czech Republic
33     Democratic Republic of the Congo
55                                 Iran
60                          Ivory Coast
68                                 Laos
80                              Moldova
91                          North Korea
104               Republic of the Congo
106                              Russia
117                         South Korea
124                               Syria
125                              Taiwan
127                            Tanzania
129                          The Gambia
137                      United Kingdom
141                           Venezuela
142                             Vietnam
Name: country_name, dtype: object

In [20]:
continents.loc[234:244]

country_name country_txt continent  \
234  United Kingdom of Great Britain and Northern I...         GBR    Europe   
235                           United States of America         USA  Americas   
236               United States Minor Outlying Islands         UMI   Oceania   
237                                            Uruguay         URY  Americas   
238                                         Uzbekistan         UZB      Asia   
239                                            Vanuatu         VUT   Oceania   
240                 Venezuela (Bolivarian Republic of)         VEN  Americas   
241                                           Viet Nam         VNM      Asia   
242                           Virgin Islands (British)         VGB  Americas   
243                              Virgin Islands (U.S.)         VIR  Americas   
244                                  Wallis and Futuna         WLF   Oceania   

                          sub-region  
234                  Northern Europe  
235                 Northern America  
236                       Micronesia  
237  Latin America and the Caribbean  
238                     Central Asia  
239                        Melanesia  
240  Latin America and the Caribbean  
241               South-eastern Asia  
242  Latin America and the Caribbean  
243  Latin America and the Caribbean  
244                        Polynesia

In [21]:
update = {
    'Bolivia (Plurinational State of)': 'Bolivia',
    'Czechia': 'Czech Republic',
    "Côte d'Ivoire": 'Ivory Coast',
    'Congo, Democratic Republic of the': 'Democratic Republic of the Congo',
    'Congo': 'Republic of the Congo',
    'Gambia': 'The Gambia',
    'Iran (Islamic Republic of)': 'Iran',
    "Korea (Democratic People's Republic of)": 'North Korea',
    'Korea, Republic of': 'South Korea',
    "Lao People's Democratic Republic": 'Laos',
    'Moldova, Republic of': 'Moldova',
    'Myanmar': 'Burma/Myanmar',
    'Russian Federation': 'Russia',
    'Syrian Arab Republic': 'Syria',
    'Taiwan, Province of China': 'Taiwan',
    'Tanzania, United Republic of': 'Tanzania',
    'Venezuela (Bolivarian Republic of)': 'Venezuela',
    'Viet Nam': 'Vietnam',
    'United Kingdom of Great Britain and Northern Ireland': 'United Kingdom'
}

In [22]:
continents.replace({"country_name": update}, inplace=True)

In [23]:
df = pd.merge(data, continents, how='left', on=['country_name'])

df.head()

country_name    pandem  pandem_dis   panback country_txt continent  \
0  Afghanistan  0.058824        0.05  0.000000         AFG      Asia   
1      Albania  0.176471        0.15  0.173110         ALB    Europe   
2      Algeria  0.294118        0.35  0.123340         DZA    Africa   
3       Angola  0.117647        0.10  0.041671         AGO    Africa   
4    Argentina  0.058824        0.05  0.054786         ARG  Americas   

                        sub-region  
0                    Southern Asia  
1                  Southern Europe  
2                  Northern Africa  
3               Sub-Saharan Africa  
4  Latin America and the Caribbean

In [24]:
df['country_txt'].isna().sum()

0

#### Adding income group according to World Bank classification

https://datahelpdesk.worldbank.org/knowledgebase/articles/906519-world-bank-country-and-lending-groups

In [25]:
classification = pd.read_excel('CLASS.xls', skiprows=4)
classification.head()

x  x.1         Economy Code    X                      Region  \
0  x    x               x    x    x                           x   
1  1  NaN     Afghanistan  AFG  NaN                  South Asia   
2  2  NaN         Albania  ALB  NaN       Europe & Central Asia   
3  3  NaN         Algeria  DZA  NaN  Middle East & North Africa   
4  4  NaN  American Samoa  ASM  NaN         East Asia & Pacific   

          Income group Lending category Other  
0                    x                x     x  
1           Low income              IDA  HIPC  
2  Upper middle income             IBRD   NaN  
3  Lower middle income             IBRD   NaN  
4  Upper middle income               ..   NaN

In [26]:
classification = classification[['Code', 'Income group']].iloc[1:219]
classification.rename(columns={"Code": "country_txt", "Income group": "income_group"}, inplace=True)
classification.tail()

country_txt         income_group
214         VIR          High income
215         PSE  Lower middle income
216         YEM           Low income
217         ZMB  Lower middle income
218         ZWE  Lower middle income

In [27]:
# Final merging
df = pd.merge(df, classification, how='left', on=['country_txt'])

In [28]:
df.to_csv('pandem.csv', index=False)

In [29]:
del df
del data

In [30]:
pandem = pd.read_csv('pandem.csv')

In [31]:
pandem.head()

country_name    pandem  pandem_dis   panback country_txt continent  \
0  Afghanistan  0.058824        0.05  0.000000         AFG      Asia   
1      Albania  0.176471        0.15  0.173110         ALB    Europe   
2      Algeria  0.294118        0.35  0.123340         DZA    Africa   
3       Angola  0.117647        0.10  0.041671         AGO    Africa   
4    Argentina  0.058824        0.05  0.054786         ARG  Americas   

                        sub-region         income_group  
0                    Southern Asia           Low income  
1                  Southern Europe  Upper middle income  
2                  Northern Africa  Lower middle income  
3               Sub-Saharan Africa  Lower middle income  
4  Latin America and the Caribbean  Upper middle income

In [32]:
pandem.rename(columns={'country_txt': 'country_text_id'}, inplace = True)

### Checking country_name compliance in vdem and pandem datasets

In [33]:
count = vdem[['country_name', 'country_text_id']]
countries_df = count.groupby('country_name')['country_text_id'].max().to_frame().reset_index().sort_values(by='country_name')

count = pandem[['country_name', 'country_text_id']]
countries_df_pandem = count.groupby('country_text_id')['country_name'].max().to_frame().reset_index().sort_values(by='country_text_id')


result = pd.merge(countries_df, countries_df_pandem, how='left', on=['country_text_id'])

result['differences'] = result['country_name_x'] == result['country_name_y']
good = list(result[result['differences'] == True].country_text_id)
missings = list(result[result['differences'] == False].country_text_id)
print(len(good))
print(pandem.shape[0])

146
146


In [34]:
vdem.rename(columns={'country_name': 'country'}, inplace = True)

#### Merging two datasets on 'country_text_id'

In [35]:
data = pd.merge(pandem, vdem, how='left', on=['country_text_id'])

In [36]:
data.head()

country_name    pandem  pandem_dis   panback country_text_id continent  \
0  Afghanistan  0.058824        0.05  0.000000             AFG      Asia   
1      Albania  0.176471        0.15  0.173110             ALB    Europe   
2      Algeria  0.294118        0.35  0.123340             DZA    Africa   
3       Angola  0.117647        0.10  0.041671             AGO    Africa   
4    Argentina  0.058824        0.05  0.054786             ARG  Americas   

                        sub-region         income_group      country  \
0                    Southern Asia           Low income  Afghanistan   
1                  Southern Europe  Upper middle income      Albania   
2                  Northern Africa  Lower middle income      Algeria   
3               Sub-Saharan Africa  Lower middle income       Angola   
4  Latin America and the Caribbean  Upper middle income    Argentina   

   rule_2017  ...  property_2019  polyarchy_2017  polyarchy_2018  \
0      0.219  ...          0.423           0.369           0.368   
1      0.514  ...          0.737           0.563           0.499   
2      0.204  ...          0.611           0.306           0.320   
3      0.306  ...          0.603           0.316           0.354   
4      0.689  ...          0.824           0.745           0.834   

   polyarchy_2019  justice_2017  justice_2018  justice_2019  education_2017  \
0           0.347         0.205         0.174         0.198           3.813   
1           0.481         0.875         0.872         0.936             NaN   
2           0.300         0.615         0.643         0.500           7.300   
3           0.386         0.451         0.451         0.487           2.980   
4           0.812         0.763         0.819         0.707           8.770   

   education_2018  education_2019  
0           3.813           3.813  
1             NaN             NaN  
2           7.300           7.300  
3           2.980           2.980  
4           8.770           8.770  

[5 rows x 26 columns]

In [37]:
data.drop(columns=['country'], inplace=True)

In [38]:
data.isna().sum()

country_name        0
pandem              0
pandem_dis          0
panback             0
country_text_id     0
continent           0
sub-region          0
income_group        0
rule_2017           0
rule_2018           0
rule_2019           0
region_pol          0
region_geo          0
property_2017       0
property_2018       0
property_2019       0
polyarchy_2017      0
polyarchy_2018      0
polyarchy_2019      0
justice_2017        0
justice_2018        0
justice_2019        0
education_2017     20
education_2018     20
education_2019     20
dtype: int64

### World Bank variables

#### Shortening dataset to variables for 2010-2019

In [39]:
names = ['Gini', 'Density', 'GDPpc', 'Inflation', 'Military', 'Oil', 'Export_GDP', 'Import_GDP', 'Mineral']
for i in range(len(names)):
    df = pd.read_csv(f'{names[i]}.csv', sep=',', skiprows=4)
    df.drop(columns = ['Indicator Code', 'Indicator Name', 'Unnamed: 64'], inplace=True)
    reshaped_df = df.melt(id_vars=['Country Name', 'Country Code'])
    reshaped_df.rename(columns={ reshaped_df.columns[0]: "country_name", reshaped_df.columns[1]: "country_text_id", reshaped_df.columns[3]: f'{names[i]}', reshaped_df.columns[2]: "year" }, inplace = True)
    reshaped_df = reshaped_df.loc[(reshaped_df.year).astype(int) >= 2010]
    if(i == 0):
        WB = reshaped_df.copy()
    else:
        WB[f'{names[i]}'] = reshaped_df[f'{names[i]}']

In [40]:
WB.head()

country_name country_text_id  year  Gini     Density         GDPpc  \
13200        Aruba             ABW  2010   NaN  564.827778  23512.602596   
13201  Afghanistan             AFG  2010   NaN   44.704082    543.303042   
13202       Angola             AGO  2010   NaN   18.734456   3587.883798   
13203      Albania             ALB  2010   NaN  106.314635   4094.350334   
13204      Andorra             AND  2010   NaN  179.678723  40852.666777   

       Inflation  Military        Oil  Export_GDP  Import_GDP   Mineral  
13200   2.078141       NaN   0.000000   61.042300   77.564852  0.000000  
13201   2.178538  1.945837   0.004974   16.182633   36.396967  0.000000  
13202 -60.496401  4.177585  38.534368   61.543113   42.580517  0.000000  
13203   3.622335  1.558592   1.499436   27.979423   48.563932  0.462026  
13204        NaN       NaN        NaN         NaN         NaN  0.000000

In [41]:
WB.isna().sum()

country_name          0
country_text_id       0
year                  0
Gini               1996
Density             301
GDPpc               169
Inflation           405
Military            897
Oil                 744
Export_GDP          392
Import_GDP          392
Mineral             659
dtype: int64

#### Filling 'missing values' with the latest value for each country

In [42]:
for i in names:
    WB[i] = WB.groupby('country_name')[i].fillna(method='ffill')

In [43]:
WB = WB.loc[(WB.year).astype(int) >= 2019]
WB.head()

country_name country_text_id  year  Gini     Density         GDPpc  \
15576        Aruba             ABW  2019   NaN  588.027778  29007.693003   
15577  Afghanistan             AFG  2019   NaN   56.937760    502.115487   
15578       Angola             AGO  2019  51.3   24.713052   2973.591160   
15579      Albania             ALB  2019  33.2  104.612263   5352.857411   
15580      Andorra             AND  2019   NaN  163.842553  40886.391165   

       Inflation  Military        Oil  Export_GDP  Import_GDP   Mineral  
15576   4.257462       NaN   0.000000   73.332115   71.632243  0.000000  
15577   2.302373  0.984561   0.001255    7.753798   37.867379  0.007421  
15578  17.145320  1.777138  15.753913   40.836290   25.541722  0.000000  
15579   1.411091  1.178901   1.365752   31.511223   45.217077  0.146426  
15580        NaN       NaN        NaN         NaN         NaN  0.000000

In [44]:
WB.isna().sum()

country_name         0
country_text_id      0
year                 0
Gini               115
Density              3
GDPpc               11
Inflation           30
Military            59
Oil                 24
Export_GDP          25
Import_GDP          25
Mineral             13
dtype: int64

#### Filling 'missing values' with 0 for 'Oil' and 'Mineral' variables (The assumption that in the absence of complete data for 2010-2019 - the country does not produce oil / mineral)

In [45]:
WB['Oil'] = WB['Oil'].fillna(0)
WB['Mineral'] = WB['Mineral'].fillna(0)

In [46]:
WB.isna().sum()

country_name         0
country_text_id      0
year                 0
Gini               115
Density              3
GDPpc               11
Inflation           30
Military            59
Oil                  0
Export_GDP          25
Import_GDP          25
Mineral              0
dtype: int64

### Checking country name compliance in both datasets

In [47]:
count = WB[['country_name', 'country_text_id']]
countries_df = count.groupby('country_name')['country_text_id'].max().to_frame().reset_index().sort_values(by='country_name')

count = pandem[['country_name', 'country_text_id']]
countries_df_pandem = count.groupby('country_text_id')['country_name'].max().to_frame().reset_index().sort_values(by='country_text_id')


result = pd.merge(countries_df_pandem, countries_df, how='left', on=['country_text_id'])

result['differences'] = result['country_name_x'] == result['country_name_y']
good = list(result[result['differences'] == True].country_text_id)
missings = list(result[result['differences'] == False].country_text_id)
print(len(good))
print(pandem.shape[0])

128
146


In [48]:
result.tail()

country_text_id country_name_x country_name_y  differences
141             VNM        Vietnam        Vietnam         True
142             YEM          Yemen    Yemen, Rep.        False
143             ZAF   South Africa   South Africa         True
144             ZMB         Zambia         Zambia         True
145             ZWE       Zimbabwe       Zimbabwe         True

In [49]:
result.isna().sum()
## Only one country without 'country_text_id' in World Bank database

country_text_id    0
country_name_x     0
country_name_y     1
differences        0
dtype: int64

In [50]:
result.loc[result['country_name_y'].isna()]

country_text_id country_name_x country_name_y  differences
133             TWN         Taiwan            NaN        False

In [51]:
WB.loc[WB.country_name.str.contains('Taiwan')]

Empty DataFrame
Columns: [country_name, country_text_id, year, Gini, Density, GDPpc, Inflation, Military, Oil, Export_GDP, Import_GDP, Mineral]
Index: []

https://datahelpdesk.worldbank.org/knowledgebase/articles/114933-where-are-your-data-on-taiwan

#### Taiwan, China, is not listed as a separate country for World Development Indicators

In [52]:
WB.drop(columns=['country_name'], inplace=True)
WB.drop(columns=['year'], inplace=True)
data = pd.merge(data, WB, how='left', on=['country_text_id'])

In [53]:
data.head()

country_name    pandem  pandem_dis   panback country_text_id continent  \
0  Afghanistan  0.058824        0.05  0.000000             AFG      Asia   
1      Albania  0.176471        0.15  0.173110             ALB    Europe   
2      Algeria  0.294118        0.35  0.123340             DZA    Africa   
3       Angola  0.117647        0.10  0.041671             AGO    Africa   
4    Argentina  0.058824        0.05  0.054786             ARG  Americas   

                        sub-region         income_group  rule_2017  rule_2018  \
0                    Southern Asia           Low income      0.219      0.254   
1                  Southern Europe  Upper middle income      0.514      0.543   
2                  Northern Africa  Lower middle income      0.204      0.200   
3               Sub-Saharan Africa  Lower middle income      0.306      0.400   
4  Latin America and the Caribbean  Upper middle income      0.689      0.732   

   ...  education_2019  Gini     Density         GDPpc  Inflation  Military  \
0  ...           3.813   NaN   56.937760    502.115487   2.302373  0.984561   
1  ...             NaN  33.2  104.612263   5352.857411   1.411091  1.178901   
2  ...           7.300  27.6   17.730075   3948.343279   1.951768  5.271414   
3  ...           2.980  51.3   24.713052   2973.591160  17.145320  1.777138   
4  ...           8.770  41.4   16.258510  10006.148974  53.548304  0.854561   

         Oil  Export_GDP  Import_GDP   Mineral  
0   0.001255    7.753798   37.867379  0.007421  
1   1.365752   31.511223   45.217077  0.146426  
2  12.313697   22.570151   29.474623  0.046717  
3  15.753913   40.836290   25.541722  0.000000  
4   0.735824   17.290218   15.103448  0.308310  

[5 rows x 34 columns]

### Fractionalization dataset

In [54]:
frac = pd.read_excel('fractionalization.xlsx')

In [55]:
frac.head()

country_name  ethnic_frac  ling_frac  relig_frac
0     Afghanistan       0.7693     0.6141      0.2717
1         Albania       0.2204     0.0400      0.4719
2         Algeria       0.3394     0.4427      0.0100
3  American Samoa       0.0000     0.1733      0.6395
4         Andorra       0.7139     0.6848      0.2326

In [56]:
countries_frac = list(frac['country_name'])
countries_pandem = list(pandem['country_name'])

In [57]:
sum([item in countries_pandem for item in countries_frac])

135

In [58]:
sum([item in countries_frac for item in countries_pandem])

135

#### 'country_name' complies in 135 cases!

#### Trial merging to check which 'country_name' differs between datasets

In [59]:
df = pd.merge(data[['country_name']], frac, how='left', on=['country_name'])

#### Checking which countries names are different in both datasets and then updating their names

In [60]:
df1 = df[df.isna().any(axis=1)]
df1['country_name']

19                        Burma/Myanmar
33     Democratic Republic of the Congo
91                          North Korea
92                      North Macedonia
104               Republic of the Congo
106                              Russia
113                            Slovakia
117                         South Korea
118                         South Sudan
129                          The Gambia
138            United States of America
Name: country_name, dtype: object

In [61]:
frac.loc[180:185]

country_name  ethnic_frac  ling_frac  relig_frac
180              Sri Lanka       0.4150     0.4645      0.4853
181  Saint Kitts and Nevis       0.1842     0.0000      0.6614
182                  Sudan       0.7147     0.7190      0.4307
183               Suriname       0.7332     0.3310      0.7910
184              Swaziland       0.0580     0.1722      0.4444
185                 Sweden       0.1000     0.1968      0.2342

In [62]:
frac.loc[frac.country_name.str.contains('Sudan')]

country_name  ethnic_frac  ling_frac  relig_frac
182        Sudan       0.7147      0.719      0.4307

#### Dictionary for updating country names (South Sudan does not exist in 2003)

In [63]:
update = {
    'Burma': 'Burma/Myanmar',
    'Democratic Republic of Congo': 'Democratic Republic of the Congo',
    'Congo': 'Republic of the Congo',
    'Gambia, The': 'The Gambia',
    "Korea, North": 'North Korea',
    'Korea, South': 'South Korea',
    'Russian Federation': 'Russia',
    'Macedonia': 'North Macedonia',
    'Slovak Republic': 'Slovakia',
    #'South Sudan'
    'United States': 'United States of America'
}

In [64]:
frac.replace({"country_name": update}, inplace=True)

### Final merging

In [65]:
data = pd.merge(data, frac, how='left', on=['country_name'])

### Full dataset with some missings

In [66]:
# data.drop(columns=['year'], inplace=True)
data['Trade_GDP'] = data['Export_GDP'] + data['Import_GDP']
data.drop(columns=['Export_GDP', 'Import_GDP'], inplace=True)
data.head()

country_name    pandem  pandem_dis   panback country_text_id continent  \
0  Afghanistan  0.058824        0.05  0.000000             AFG      Asia   
1      Albania  0.176471        0.15  0.173110             ALB    Europe   
2      Algeria  0.294118        0.35  0.123340             DZA    Africa   
3       Angola  0.117647        0.10  0.041671             AGO    Africa   
4    Argentina  0.058824        0.05  0.054786             ARG  Americas   

                        sub-region         income_group  rule_2017  rule_2018  \
0                    Southern Asia           Low income      0.219      0.254   
1                  Southern Europe  Upper middle income      0.514      0.543   
2                  Northern Africa  Lower middle income      0.204      0.200   
3               Sub-Saharan Africa  Lower middle income      0.306      0.400   
4  Latin America and the Caribbean  Upper middle income      0.689      0.732   

   ...     Density         GDPpc  Inflation  Military        Oil   Mineral  \
0  ...   56.937760    502.115487   2.302373  0.984561   0.001255  0.007421   
1  ...  104.612263   5352.857411   1.411091  1.178901   1.365752  0.146426   
2  ...   17.730075   3948.343279   1.951768  5.271414  12.313697  0.046717   
3  ...   24.713052   2973.591160  17.145320  1.777138  15.753913  0.000000   
4  ...   16.258510  10006.148974  53.548304  0.854561   0.735824  0.308310   

   ethnic_frac  ling_frac  relig_frac  Trade_GDP  
0       0.7693     0.6141      0.2717  45.621177  
1       0.2204     0.0400      0.4719  76.728300  
2       0.3394     0.4427      0.0100  52.044775  
3       0.7867     0.7870      0.6276  66.378012  
4       0.2550     0.0620      0.2236  32.393666  

[5 rows x 36 columns]

### Missing values

In [67]:
data.isna().sum()

country_name        0
pandem              0
pandem_dis          0
panback             0
country_text_id     0
continent           0
sub-region          0
income_group        0
rule_2017           0
rule_2018           0
rule_2019           0
region_pol          0
region_geo          0
property_2017       0
property_2018       0
property_2019       0
polyarchy_2017      0
polyarchy_2018      0
polyarchy_2019      0
justice_2017        0
justice_2018        0
justice_2019        0
education_2017     20
education_2018     20
education_2019     20
Gini               25
Density             3
GDPpc               4
Inflation           8
Military            5
Oil                 1
Mineral             1
ethnic_frac         1
ling_frac           1
relig_frac          1
Trade_GDP           6
dtype: int64

### Data imputation

#### Fractionalization for South Sudan
In 2003 South Sudan was a part of Sudan, so we impute for South Sudan values for Sudan

In [68]:
data.loc[data['country_name'] == 'South Sudan','ethnic_frac'] = data.loc[data['country_name'] == 'Sudan']['ethnic_frac'].values[0]
data.loc[data['country_name'] == 'South Sudan','ling_frac'] = data.loc[data['country_name'] == 'Sudan']['ling_frac'].values[0]
data.loc[data['country_name'] == 'South Sudan','relig_frac'] = data.loc[data['country_name'] == 'Sudan']['relig_frac'].values[0]

In [69]:
data.loc[data['country_name'] == 'South Sudan']

country_name    pandem  pandem_dis  panback country_text_id continent  \
118  South Sudan  0.176471        0.15      0.0             SSD    Africa   

             sub-region income_group  rule_2017  rule_2018  ...  Density  \
118  Sub-Saharan Africa   Low income      0.074      0.051  ...      NaN   

           GDPpc   Inflation  Military        Oil  Mineral  ethnic_frac  \
118  1119.651333  187.851626  1.329208  31.325699      0.0       0.7147   

     ling_frac  relig_frac  Trade_GDP  
118      0.719      0.4307  65.551349  

[1 rows x 36 columns]

In [70]:
data.loc[data['relig_frac'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, rule_2017, rule_2018, rule_2019, region_pol, region_geo, property_2017, property_2018, property_2019, polyarchy_2017, polyarchy_2018, polyarchy_2019, justice_2017, justice_2018, justice_2019, education_2017, education_2018, education_2019, Gini, Density, GDPpc, Inflation, Military, Oil, Mineral, ethnic_frac, ling_frac, relig_frac, Trade_GDP]
Index: []

[0 rows x 36 columns]

### 'Oil' and 'Mineral' for Taiwan
"Taiwan produces a small amount of natural gas." We impute 0 for Oil and Mineral for simplicity.

In [71]:
data.loc[data['country_name'] == 'Taiwan']

country_name    pandem  pandem_dis   panback country_text_id continent  \
125       Taiwan  0.058824        0.05  0.049692             TWN      Asia   

       sub-region income_group  rule_2017  rule_2018  ...  Density  GDPpc  \
125  Eastern Asia  High income      0.911       0.92  ...      NaN    NaN   

     Inflation  Military  Oil  Mineral  ethnic_frac  ling_frac  relig_frac  \
125        NaN       NaN  NaN      NaN       0.2744     0.5028      0.6845   

     Trade_GDP  
125        NaN  

[1 rows x 36 columns]

In [72]:
data.loc[data['country_name'] == 'Taiwan','Oil'] = 0
data.loc[data['country_name'] == 'Taiwan','Mineral'] = 0

In [73]:
data.loc[data['country_name'] == 'Taiwan']

country_name    pandem  pandem_dis   panback country_text_id continent  \
125       Taiwan  0.058824        0.05  0.049692             TWN      Asia   

       sub-region income_group  rule_2017  rule_2018  ...  Density  GDPpc  \
125  Eastern Asia  High income      0.911       0.92  ...      NaN    NaN   

     Inflation  Military  Oil  Mineral  ethnic_frac  ling_frac  relig_frac  \
125        NaN       NaN  0.0      0.0       0.2744     0.5028      0.6845   

     Trade_GDP  
125        NaN  

[1 rows x 36 columns]

### Imputation of 'Density'.

Wikipedia as a source.

In [74]:
data.loc[data['Density'].isna()]

country_name    pandem  pandem_dis   panback country_text_id continent  \
118  South Sudan  0.176471        0.15  0.000000             SSD    Africa   
121        Sudan  0.176471        0.15  0.056069             SDN    Africa   
125       Taiwan  0.058824        0.05  0.049692             TWN      Asia   

             sub-region income_group  rule_2017  rule_2018  ...  Density  \
118  Sub-Saharan Africa   Low income      0.074      0.051  ...      NaN   
121     Northern Africa   Low income      0.185      0.167  ...      NaN   
125        Eastern Asia  High income      0.911      0.920  ...      NaN   

           GDPpc   Inflation  Military        Oil   Mineral  ethnic_frac  \
118  1119.651333  187.851626  1.329208  31.325699  0.000000       0.7147   
121   441.505603   50.994052  2.278302   1.013017  2.537048       0.7147   
125          NaN         NaN       NaN   0.000000  0.000000       0.2744   

     ling_frac  relig_frac  Trade_GDP  
118     0.7190      0.4307  65.551349  
121     0.7190      0.4307  26.721863  
125     0.5028      0.6845        NaN  

[3 rows x 36 columns]

In [75]:
data.loc[data['country_name'] == 'Taiwan','Density'] = 650
data.loc[data['country_name'] == 'Sudan','Density'] = 21.3
data.loc[data['country_name'] == 'South Sudan','Density'] = 13.33

In [76]:
data.loc[data['Density'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, rule_2017, rule_2018, rule_2019, region_pol, region_geo, property_2017, property_2018, property_2019, polyarchy_2017, polyarchy_2018, polyarchy_2019, justice_2017, justice_2018, justice_2019, education_2017, education_2018, education_2019, Gini, Density, GDPpc, Inflation, Military, Oil, Mineral, ethnic_frac, ling_frac, relig_frac, Trade_GDP]
Index: []

[0 rows x 36 columns]

In [77]:
data.isna().sum()

country_name        0
pandem              0
pandem_dis          0
panback             0
country_text_id     0
continent           0
sub-region          0
income_group        0
rule_2017           0
rule_2018           0
rule_2019           0
region_pol          0
region_geo          0
property_2017       0
property_2018       0
property_2019       0
polyarchy_2017      0
polyarchy_2018      0
polyarchy_2019      0
justice_2017        0
justice_2018        0
justice_2019        0
education_2017     20
education_2018     20
education_2019     20
Gini               25
Density             0
GDPpc               4
Inflation           8
Military            5
Oil                 0
Mineral             0
ethnic_frac         0
ling_frac           0
relig_frac          0
Trade_GDP           6
dtype: int64

### Imputation of 'GDPpc'

Wikipedia as a source

In [78]:
data.loc[data['GDPpc'].isna()]

country_name    pandem  pandem_dis   panback country_text_id continent  \
91   North Korea  0.176471        0.15  0.000000             PRK      Asia   
115      Somalia  0.352941        0.30  0.129312             SOM    Africa   
124        Syria  0.235294        0.25  0.038730             SYR      Asia   
125       Taiwan  0.058824        0.05  0.049692             TWN      Asia   

             sub-region income_group  rule_2017  rule_2018  ...     Density  \
91         Eastern Asia   Low income      0.032      0.041  ...  212.190175   
115  Sub-Saharan Africa   Low income      0.142      0.103  ...   23.923477   
124        Western Asia   Low income      0.058      0.058  ...   92.067108   
125        Eastern Asia  High income      0.911      0.920  ...  650.000000   

     GDPpc  Inflation  Military  Oil  Mineral  ethnic_frac  ling_frac  \
91     NaN        NaN       NaN  0.0      0.0       0.0000     0.0000   
115    NaN        NaN       NaN  0.0      0.0       0.8117     0.0330   
124    NaN  36.702295  4.056142  0.0      0.0       0.5399     0.1817   
125    NaN        NaN       NaN  0.0      0.0       0.2744     0.5028   

     relig_frac  Trade_GDP  
91       0.6604        NaN  
115      0.0000        NaN  
124      0.4310        NaN  
125      0.6845        NaN  

[4 rows x 36 columns]

In [79]:
data.loc[data['country_name'] == 'North Korea','GDPpc'] = 1300
data.loc[data['country_name'] == 'Somalia','GDPpc'] = 348
data.loc[data['country_name'] == 'Syria','GDPpc'] = 831
data.loc[data['country_name'] == 'Taiwan','GDPpc'] = 24828

In [80]:
data.loc[data['GDPpc'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, rule_2017, rule_2018, rule_2019, region_pol, region_geo, property_2017, property_2018, property_2019, polyarchy_2017, polyarchy_2018, polyarchy_2019, justice_2017, justice_2018, justice_2019, education_2017, education_2018, education_2019, Gini, Density, GDPpc, Inflation, Military, Oil, Mineral, ethnic_frac, ling_frac, relig_frac, Trade_GDP]
Index: []

[0 rows x 36 columns]

In [81]:
data.isna().sum()

country_name        0
pandem              0
pandem_dis          0
panback             0
country_text_id     0
continent           0
sub-region          0
income_group        0
rule_2017           0
rule_2018           0
rule_2019           0
region_pol          0
region_geo          0
property_2017       0
property_2018       0
property_2019       0
polyarchy_2017      0
polyarchy_2018      0
polyarchy_2019      0
justice_2017        0
justice_2018        0
justice_2019        0
education_2017     20
education_2018     20
education_2019     20
Gini               25
Density             0
GDPpc               0
Inflation           8
Military            5
Oil                 0
Mineral             0
ethnic_frac         0
ling_frac           0
relig_frac          0
Trade_GDP           6
dtype: int64

### Imputation of 'Gini'
* from Wikipedia
* mean according to income group for unavailable data

In [82]:
data.loc[data['Gini'].isna()]['country_name'].to_list()

['Afghanistan',
 'Azerbaijan',
 'Cambodia',
 'Central African Republic',
 'Cuba',
 'Eritrea',
 'Jamaica',
 'Kuwait',
 'Libya',
 'Mali',
 'New Zealand',
 'Nigeria',
 'North Korea',
 'Oman',
 'Papua New Guinea',
 'Qatar',
 'Saudi Arabia',
 'Singapore',
 'Somalia',
 'South Sudan',
 'Syria',
 'Taiwan',
 'Turkmenistan',
 'Uzbekistan',
 'Venezuela']

In [83]:
### From Wikipedia:
data.loc[data['country_name'] == 'Afghanistan','Gini'] = 27.8
data.loc[data['country_name'] == 'Azerbaijan','Gini'] = 26.6
data.loc[data['country_name'] == 'Cambodia','Gini'] = 36
data.loc[data['country_name'] == 'Central African Republic','Gini'] = 56.3
data.loc[data['country_name'] == 'Cuba','Gini'] = 38
data.loc[data['country_name'] == 'Jamaica','Gini'] = 35
data.loc[data['country_name'] == 'Mali','Gini'] = 33
data.loc[data['country_name'] == 'New Zealand','Gini'] = 33.9
data.loc[data['country_name'] == 'Nigeria','Gini'] = 35.1
data.loc[data['country_name'] == 'Papua New Guinea','Gini'] = 41.9
data.loc[data['country_name'] == 'Qatar','Gini'] = 41.1
data.loc[data['country_name'] == 'Saudi Arabia','Gini'] = 45.9
data.loc[data['country_name'] == 'Singapore','Gini'] = 45.9
data.loc[data['country_name'] == 'South Sudan','Gini'] = 45.5
data.loc[data['country_name'] == 'Syria','Gini'] = 55.8
data.loc[data['country_name'] == 'Taiwan','Gini'] = 34.1
data.loc[data['country_name'] == 'Turkmenistan','Gini'] = 40.8
data.loc[data['country_name'] == 'Uzbekistan','Gini'] = 36.7
data.loc[data['country_name'] == 'Venezuela','Gini'] = 44.8

#### Countries with missings

In [84]:
to_fill = data.loc[data['Gini'].isna()]['country_name'].to_list()
to_fill

['Eritrea', 'Kuwait', 'Libya', 'North Korea', 'Oman', 'Somalia']

Filling NaN's for 'Gini' with means for income group to which country belongs

In [85]:
data.groupby('income_group')['Gini'].mean()

income_group
High income            33.800000
Low income             40.180000
Lower middle income    38.842105
Upper middle income    40.386842
Name: Gini, dtype: float64

In [86]:
high = data.groupby('income_group')['Gini'].mean()['High income']
low = data.groupby('income_group')['Gini'].mean()['Low income']
lmid = data.groupby('income_group')['Gini'].mean()['Lower middle income']
umid = data.groupby('income_group')['Gini'].mean()['Upper middle income']

In [87]:
for i in to_fill:
    if(data.loc[data['country_name'] == i, 'income_group'].values[0] == 'High income'):
        data.loc[data['country_name'] == i,'Gini'] = high
    if(data.loc[data['country_name'] == i, 'income_group'].values[0] == 'Low income'):
        data.loc[data['country_name'] == i,'Gini'] = low  
    if(data.loc[data['country_name'] == i, 'income_group'].values[0] == 'Lower middle income'):
        data.loc[data['country_name'] == i,'Gini'] = lmid
    else:
        data.loc[data['country_name'] == i,'Gini'] = umid

In [88]:
data.isna().sum()

country_name        0
pandem              0
pandem_dis          0
panback             0
country_text_id     0
continent           0
sub-region          0
income_group        0
rule_2017           0
rule_2018           0
rule_2019           0
region_pol          0
region_geo          0
property_2017       0
property_2018       0
property_2019       0
polyarchy_2017      0
polyarchy_2018      0
polyarchy_2019      0
justice_2017        0
justice_2018        0
justice_2019        0
education_2017     20
education_2018     20
education_2019     20
Gini                0
Density             0
GDPpc               0
Inflation           8
Military            5
Oil                 0
Mineral             0
ethnic_frac         0
ling_frac           0
relig_frac          0
Trade_GDP           6
dtype: int64

### Imputation of other variables
* mean according to income group

##### Function for imputing missings for concrete variable with means for income group to which country belongs

In [89]:
def mean_filler(var):
    
    to_fill = data.loc[data[var].isna()]['country_name'].to_list()

    high = data.groupby('income_group')[var].mean()['High income']
    low = data.groupby('income_group')[var].mean()['Low income']
    lmid = data.groupby('income_group')[var].mean()['Lower middle income']
    umid = data.groupby('income_group')[var].mean()['Upper middle income']

    for i in to_fill:
        if(data.loc[data['country_name'] == i, 'income_group'].values[0] == 'High income'):
            data.loc[data['country_name'] == i,var] = high
        if(data.loc[data['country_name'] == i, 'income_group'].values[0] == 'Low income'):
            data.loc[data['country_name'] == i,var] = low  
        if(data.loc[data['country_name'] == i, 'income_group'].values[0] == 'Lower middle income'):
            data.loc[data['country_name'] == i,var] = lmid
        else:
            data.loc[data['country_name'] == i,var] = umid

In [90]:
data.loc[data['education_2019'].isna()]['country_name'].to_list()

['Albania',
 'Bosnia and Herzegovina',
 'Croatia',
 'Eritrea',
 'Ethiopia',
 'Indonesia',
 'Kuwait',
 'Mongolia',
 'North Macedonia',
 'Oman',
 'Papua New Guinea',
 'Qatar',
 'Serbia',
 'Slovenia',
 'South Sudan',
 'Sudan',
 'Taiwan',
 'Turkmenistan',
 'United Arab Emirates',
 'Yemen']

In [91]:
mean_filler('education_2017')
data.loc[data['education_2017'].isna()]
mean_filler('education_2018')
data.loc[data['education_2018'].isna()]
mean_filler('education_2019')
data.loc[data['education_2019'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, rule_2017, rule_2018, rule_2019, region_pol, region_geo, property_2017, property_2018, property_2019, polyarchy_2017, polyarchy_2018, polyarchy_2019, justice_2017, justice_2018, justice_2019, education_2017, education_2018, education_2019, Gini, Density, GDPpc, Inflation, Military, Oil, Mineral, ethnic_frac, ling_frac, relig_frac, Trade_GDP]
Index: []

[0 rows x 36 columns]

In [92]:
data.loc[data['Inflation'].isna()]['country_name'].to_list()

['Central African Republic',
 'Cuba',
 'Eritrea',
 'North Korea',
 'Somalia',
 'Taiwan',
 'Turkmenistan',
 'Uzbekistan']

In [93]:
mean_filler('Inflation')

In [94]:
data.loc[data['Inflation'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, rule_2017, rule_2018, rule_2019, region_pol, region_geo, property_2017, property_2018, property_2019, polyarchy_2017, polyarchy_2018, polyarchy_2019, justice_2017, justice_2018, justice_2019, education_2017, education_2018, education_2019, Gini, Density, GDPpc, Inflation, Military, Oil, Mineral, ethnic_frac, ling_frac, relig_frac, Trade_GDP]
Index: []

[0 rows x 36 columns]

In [95]:
data.loc[data['Military'].isna()]['country_name'].to_list()

['Eritrea', 'North Korea', 'Somalia', 'Taiwan', 'Turkmenistan']

In [96]:
mean_filler('Military')

In [97]:
data.loc[data['Military'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, rule_2017, rule_2018, rule_2019, region_pol, region_geo, property_2017, property_2018, property_2019, polyarchy_2017, polyarchy_2018, polyarchy_2019, justice_2017, justice_2018, justice_2019, education_2017, education_2018, education_2019, Gini, Density, GDPpc, Inflation, Military, Oil, Mineral, ethnic_frac, ling_frac, relig_frac, Trade_GDP]
Index: []

[0 rows x 36 columns]

In [98]:
data.loc[data['Trade_GDP'].isna()]['country_name'].to_list()

['North Korea', 'Papua New Guinea', 'Somalia', 'Syria', 'Taiwan', 'Yemen']

In [99]:
mean_filler('Trade_GDP')

In [100]:
data.loc[data['Trade_GDP'].isna()]

Empty DataFrame
Columns: [country_name, pandem, pandem_dis, panback, country_text_id, continent, sub-region, income_group, rule_2017, rule_2018, rule_2019, region_pol, region_geo, property_2017, property_2018, property_2019, polyarchy_2017, polyarchy_2018, polyarchy_2019, justice_2017, justice_2018, justice_2019, education_2017, education_2018, education_2019, Gini, Density, GDPpc, Inflation, Military, Oil, Mineral, ethnic_frac, ling_frac, relig_frac, Trade_GDP]
Index: []

[0 rows x 36 columns]

In [101]:
data.isna().sum()

country_name       0
pandem             0
pandem_dis         0
panback            0
country_text_id    0
continent          0
sub-region         0
income_group       0
rule_2017          0
rule_2018          0
rule_2019          0
region_pol         0
region_geo         0
property_2017      0
property_2018      0
property_2019      0
polyarchy_2017     0
polyarchy_2018     0
polyarchy_2019     0
justice_2017       0
justice_2018       0
justice_2019       0
education_2017     0
education_2018     0
education_2019     0
Gini               0
Density            0
GDPpc              0
Inflation          0
Military           0
Oil                0
Mineral            0
ethnic_frac        0
ling_frac          0
relig_frac         0
Trade_GDP          0
dtype: int64

In [102]:
changes = ['rule', 'property', 'polyarchy', 'justice', 'education']

for i in changes:
    data[f'{i}_pct_change_2017'] = (data[f'{i}_2019'] - data[f'{i}_2017']) / data[f'{i}_2017']
    data[f'{i}_pct_change_2018'] = (data[f'{i}_2019'] - data[f'{i}_2018']) / data[f'{i}_2018']

In [103]:
data.rename(columns={ 'rule_2019': "rule", 'property_2019': "property", 'polyarchy_2019': "polyarchy", 'justice_2019': "justice", 'education_2019': "education"}, inplace = True)
data.drop(columns=['rule_2017', 'rule_2018', 'property_2017', 'property_2018', 'polyarchy_2017', 'polyarchy_2018', 'justice_2017', 'justice_2018', 'education_2017', 'education_2018'], inplace=True)

In [104]:
data.head()

country_name    pandem  pandem_dis   panback country_text_id continent  \
0  Afghanistan  0.058824        0.05  0.000000             AFG      Asia   
1      Albania  0.176471        0.15  0.173110             ALB    Europe   
2      Algeria  0.294118        0.35  0.123340             DZA    Africa   
3       Angola  0.117647        0.10  0.041671             AGO    Africa   
4    Argentina  0.058824        0.05  0.054786             ARG  Americas   

                        sub-region         income_group   rule  region_pol  \
0                    Southern Asia           Low income  0.341           8   
1                  Southern Europe  Upper middle income  0.596           1   
2                  Northern Africa  Lower middle income  0.217           3   
3               Sub-Saharan Africa  Lower middle income  0.406           4   
4  Latin America and the Caribbean  Upper middle income  0.757           2   

   ...  rule_pct_change_2017  rule_pct_change_2018  property_pct_change_2017  \
0  ...              0.557078              0.342520                  0.014388   
1  ...              0.159533              0.097606                 -0.023841   
2  ...              0.063725              0.085000                 -0.154910   
3  ...              0.326797              0.015000                 -0.017915   
4  ...              0.098694              0.034153                 -0.055046   

   property_pct_change_2018  polyarchy_pct_change_2017  \
0                 -0.029817                  -0.059621   
1                 -0.021248                  -0.145648   
2                 -0.161866                  -0.019608   
3                  0.039655                   0.221519   
4                 -0.050691                   0.089933   

   polyarchy_pct_change_2018  justice_pct_change_2017  \
0                  -0.057065                -0.034146   
1                  -0.036072                 0.069714   
2                  -0.062500                -0.186992   
3                   0.090395                 0.079823   
4                  -0.026379                -0.073394   

   justice_pct_change_2018  education_pct_change_2017  \
0                 0.137931                        0.0   
1                 0.073394                        0.0   
2                -0.222395                        0.0   
3                 0.079823                        0.0   
4                -0.136752                        0.0   

   education_pct_change_2018  
0                        0.0  
1                        0.0  
2                        0.0  
3                        0.0  
4                        0.0  

[5 rows x 36 columns]

In [105]:
cols = ['rule_pct_change_2017',
 'rule_pct_change_2018',
 'property_pct_change_2017',
 'property_pct_change_2018',
 'polyarchy_pct_change_2017',
 'polyarchy_pct_change_2018',
 'justice_pct_change_2017',
 'justice_pct_change_2018',
 'education_pct_change_2017',
 'education_pct_change_2018']

for i in range(len(cols)):
    print(cols[i])
    print(data[cols[i]].nunique())

rule_pct_change_2017
142
rule_pct_change_2018
139
property_pct_change_2017
137
property_pct_change_2018
119
polyarchy_pct_change_2017
146
polyarchy_pct_change_2018
141
justice_pct_change_2017
142
justice_pct_change_2018
125
education_pct_change_2017
1
education_pct_change_2018
1


In [106]:
data.drop(columns=['education_pct_change_2017', 'education_pct_change_2018'], inplace=True)

In [107]:
data.columns.to_list()

['country_name',
 'pandem',
 'pandem_dis',
 'panback',
 'country_text_id',
 'continent',
 'sub-region',
 'income_group',
 'rule',
 'region_pol',
 'region_geo',
 'property',
 'polyarchy',
 'justice',
 'education',
 'Gini',
 'Density',
 'GDPpc',
 'Inflation',
 'Military',
 'Oil',
 'Mineral',
 'ethnic_frac',
 'ling_frac',
 'relig_frac',
 'Trade_GDP',
 'rule_pct_change_2017',
 'rule_pct_change_2018',
 'property_pct_change_2017',
 'property_pct_change_2018',
 'polyarchy_pct_change_2017',
 'polyarchy_pct_change_2018',
 'justice_pct_change_2017',
 'justice_pct_change_2018']

In [108]:
data.rename(columns={'sub-region': 'sub_region'}, inplace = True)
data = data[[
    'country_name','country_text_id', # country name/id
    'continent', 'sub_region', 'region_pol', 'region_geo', 'income_group', # classification groups
    'pandem', 'pandem_dis', 'panback', # pandemic data
    'rule', 'property', 'justice', # rule of law (vdem)
    'polyarchy', 'education', # other variables (vdem)
    'rule_pct_change_2017', 'rule_pct_change_2018', 'property_pct_change_2017', 'property_pct_change_2018', 
    'polyarchy_pct_change_2017', 'polyarchy_pct_change_2018', 'justice_pct_change_2017', 'justice_pct_change_2018', # percentage changes
    'Gini', 'Density', 'GDPpc', 'Inflation', 'Military', 'Oil', 'Mineral', 'Trade_GDP', # World Bank variables
    'ethnic_frac', 'ling_frac', 'relig_frac' # Fractionalization variables
 ]]

In [109]:
data.to_csv('final_dataset.csv', index=False)

In [110]:
data.head()

country_name country_text_id continent                       sub_region  \
0  Afghanistan             AFG      Asia                    Southern Asia   
1      Albania             ALB    Europe                  Southern Europe   
2      Algeria             DZA    Africa                  Northern Africa   
3       Angola             AGO    Africa               Sub-Saharan Africa   
4    Argentina             ARG  Americas  Latin America and the Caribbean   

   region_pol  region_geo         income_group    pandem  pandem_dis  \
0           8          14           Low income  0.058824        0.05   
1           1           3  Upper middle income  0.176471        0.15   
2           3           5  Lower middle income  0.294118        0.35   
3           4           7  Lower middle income  0.117647        0.10   
4           2          18  Upper middle income  0.058824        0.05   

    panback  ...     Density         GDPpc  Inflation  Military        Oil  \
0  0.000000  ...   56.937760    502.115487   2.302373  0.984561   0.001255   
1  0.173110  ...  104.612263   5352.857411   1.411091  1.178901   1.365752   
2  0.123340  ...   17.730075   3948.343279   1.951768  5.271414  12.313697   
3  0.041671  ...   24.713052   2973.591160  17.145320  1.777138  15.753913   
4  0.054786  ...   16.258510  10006.148974  53.548304  0.854561   0.735824   

    Mineral  Trade_GDP  ethnic_frac  ling_frac  relig_frac  
0  0.007421  45.621177       0.7693     0.6141      0.2717  
1  0.146426  76.728300       0.2204     0.0400      0.4719  
2  0.046717  52.044775       0.3394     0.4427      0.0100  
3  0.000000  66.378012       0.7867     0.7870      0.6276  
4  0.308310  32.393666       0.2550     0.0620      0.2236  

[5 rows x 34 columns]

In [111]:
data.shape

(146, 34)

### Variables description:
##### Country name/id and location:
* 'country_name'
* 'country_text_id'
* 'continent' - 5 categories ('Asia', 'Europe', 'Africa', 'Americas', 'Oceania')
* 'sub-region' - 15 categories (more advanced division than in the case of continents - e.g. 'Northern Europe')

##### V-dem covid data:
* 'pandem' - is the sum of restrictions (6 types of violations - with 4 levels (0 for non, 1 for minor, 2 for some, 3 for major violations) divided by all violations recorded (maximum - 17). In our case pandem ranges from 0 to 9 (maximum: 9 out of 17 violations = 0.5294).
* 'pandem_dis' - 'pandem_dis' is builded on the 'pandem' index. 'pandem_dis' variable is the sum of restrictions (6 types of violations - with 4 levels (0 for non, 1 for minor, 2 for some, 3 for major violations) + government disinformation indicator ('govmis') with 3 levels (0 for never, 1 for occasionally, 2 for often, 3 for always or almost always) divided by the total additive score (maximum - 20 (17 for pandem and 3 for govmis)). In our case 'pandem_dis' ranges from 0 to 10 (maximum: 10 out of 20 = 0.5).
* 'panback' - is builded on the 'pandem_alt' index (which is the alternative version of 'pandem' index) and 'v2x_libdem' index (Liberal democracy index) from 2019 year for each country. In our case 'panback' has 112 unique values, so we treat 'panback' as a fully continuous variable. "The Panback index captures such inverted U-shaped vulnerability to backsliding by weighing more heavily violations recorded in countries in the mid-range of V-Dem's liberal democracy index".


##### World Bank classification (income group)
* 'income_group' - according to World Bank classification, allocation countries to one of the four groups: 'Low income', 'Upper middle income', 'Lower middle income' and 'High income'.

##### V-dem dataset:
* 'rule' ('v2x_rule' in V-dem) - "To what extent are laws transparently, independently, predictably, impartially, and equally enforced, and to what extent do the actions of government officials comply with the law?". Interval, from low to high (0-1).
* 'justice' ('v2xcl_acjst' in V-dem) - "Do citizens enjoy secure and effective access to justice?". Interval, from low to high (0-1).
* 'property' ('v2xcl_prpty' in V-dem) - "Do citizens enjoy the right to private property?". Interval, from low to high (0-1).
* 'polyarchy' ('v2x_polyarchy' in V-dem) - Electoral democracy index. "To what extent is the ideal of electoral democracy in its fullest sense achieved?". Interval, from low to high (0-1).
* 'education' ('e_peaveduc' in V-dem) - "What is the average years of education among citizens older than 15?".
* 'region_pol' ('e_regionpol' in V-dem) - "In which politico-geographic region is this country located?". (8 levels: 1. Eastern Europe and post Soviet Union (including Central Asia, Mongolia); 2. Latin America; 3. North Africa and the Middle East; 4. Sub–Saharan Africa; 5. Western Europe and North America (including Australia, New Zealand); 6. East Asia (including Japan, excluding Mongolia); 7. South–East Asia; 8. South Asia.)
* 'region_geo' ('e_regiongeo' in V-dem) - "In which geographic region is this country located?". (19 levels).


##### World Bank database:
* 'Gini' - "Gini index measures the extent to which the distribution of income (or, in some cases, consumption expenditure) among individuals or households within an economy deviates from a perfectly equal distribution."
* 'Density' - "Population density is midyear population divided by land area in square kilometers."
* 'GDPpc' - "GDP per capita is gross domestic product divided by midyear population. Data are in current U.S. dollars."
* 'Inflation' - "Inflation as measured by the consumer price index reflects the annual percentage change in the cost to the average consumer of acquiring a basket of goods and services that may be fixed or changed at specified intervals, such as yearly. The Laspeyres formula is generally used."
* 'Military' - military expenditures data (% of GDP).
* 'Oil' - "Oil rents are the difference between the value of crude oil production at world prices and total costs of production." (% of GDP).
* 'Mineral' - "Mineral rents are the difference between the value of production for a stock of minerals at world prices and their total costs of production. Minerals included in the calculation are tin, gold, lead, zinc, iron, copper, nickel, silver, bauxite, and phosphate." (% of GDP).
* 'Trade_GDP' - sum of exports and imports of goods and services (% of GDP).


##### Fractionalization data (based on Alesina et al's analysis, 2003):
Fractionalization measures are computing the probability that two randomly drawn individuals are not from the same group :
* 'ethnic_frac' - ethnic fractionalization
* 'ling_frac' - linguistic fractionalization
* 'relig_frac' - religious fractionalization